### PR TITLE
Add RunWithSynchronizationContext with arguments

### LINKED ISSUE #916: RunWithSynchronizationContext With Args
Can you provide a RunWithSynchronizationContext method that accepts args to pass to Application similar to regular Run? Needed so we can still use our dbus service...

### PR DESCRIPTION
**Pull request by fsobolev**
- [x] I agree that my contribution may be licensed either under MIT or any version of LGPL license.

Closes #916

### CODE REVIEW

File: src/Libs/Gio-2.0/Public/Application.cs [hunk 1]
```diff
@@ -14,19 +14,21 @@ public int Run()
...
-            return Run();
+            return Run(argc, argv);
         }
         finally
         {
             SynchronizationContext.SetSynchronizationContext(original);
         }
     }
+
+    public int RunWithSynchronizationContext() => RunWithSynchronizationContext(0, new string[0]);
```
Comment (nlogozzo) [inline src/Libs/Gio-2.0/Public/Application.cs:33]: `Array.Empty<string>()` would be better than `new string[0]`

**Review by nlogozzo — CHANGES_REQUESTED** [review]

File: src/Libs/Gio-2.0/Public/Application.cs (hunk 1 above)
Comment (fsobolev) [reply src/Libs/Gio-2.0/Public/Application.cs:33]: @badcel should this be changed on line 14 as well?

**Comment by badcel** [comment]
I think we should remove the parameterless method (`Run` and `RunWithSynchronizationContext`) as it is not available in the bindings, too. So let's keep the additions to a minimum.

Unfortunately the generated bindings are currently not 100% correct.

Meaning the encoding is probably not correct. I would ignore this for now as there are no length based platform arrays currently and it looks like it works more or less. (Should be fixed with #955.)

Additionally the length parameter is part of the public API. As it is declared to belong to the array it will vanish if #917 is merged. For this reason I would remove it already from the  `RunWithSynchronizationContext` to avoid breaking the API again.

The last point is that the array itself is nullable. For this reason I think we should mark it as nullable, too (see #955).

If it is null you can simply pass `0` and `Array.Empty<string>` to the current run implementation.

(See [documentation](https://docs.gtk.org/gio/method.Application.run.html)

After updating the code: Can you confirm that it works for you with `Array.Empty`, `null` and actual data as a parameter?

**Comment by fsobolev** [comment]
Unfortunately there are some problems, but I guess it's not related to this PR?

| Arguments \ Flag | DefaultFlags | HandlesOpen | HandlesCommandLine |
| --- | --- | --- | --- |
| `null` | ✅ Calls `OnActivate` |  ✅ Calls `OnActivate` | ⚠️ <sup>[2]</sup> |
| `Array.Empty<string>()` | ✅ Calls `OnActivate`  | ✅ Calls `OnActivate` | ⚠️ <sup>[2]</sup> |
| `new string[] { ... }` | ✅ Closes with `This application can not open files.` | ⚠️ <sup>[1]</sup> | ⚠️ <sup>[2]</sup> |

<sup>[1]</sup> Calls `OnOpen` successfully, but trying to get the files (`e.Files`) causes crash:
```
UnhandledException - unhandled exception: System.InvalidCastException: Unable to cast object of type 'System.IntPtr' to type 'Gio.File[]'.
   at GObject.Value.Extract[T]() in /home/user/.var/distrobox/fedora/Projects/gir.core/src/Libs/GObject-2.0/Public/Value.cs:line 167
   at Gio.Application.OpenSignalArgs.get_Files() in /home/user/.var/distrobox/fedora/Projects/gir.core/src/Libs/Gio-2.0/Public/Application.Signals.Generated.cs:line 114
   at org.gircore.test.Program.<>c.<.ctor>b__3_3(Application sender, OpenSignalArgs e) in /home/user/.var/distrobox/fedora/Projects/GirTest/Program.cs:line 35
   at GObject.Signal`2.<>c__DisplayClass11_0.<Connect>b__0(Value returnValue, Value[] parameters) in /home/user/.var/distrobox/fedora/Projects/gir.core/src/Libs/GObject-2.0/Public/SignalTSenderTSignalArgs.cs:line 46
   at GObject.Closure.InternalCallback(IntPtr closure, IntPtr returnValuePtr, UInt32 nParamValues, ValueData[] paramValuesData, IntPtr invocationHint, IntPtr userData) in /home/user/.var/distrobox/fedora/Projects/gir.core/src/Libs/GObject-2.0/Public/Closure.cs:line 47
```

<sup>[2]</sup> Calls `OnCommandLine` succesfully, but trying to get arguments (`var argv = e.CommandLine.GetArguments(out var argc);`) causes crash:
```
UnhandledException - unhandled exception: System.Runtime.InteropServices.MarshalDirectiveException: Cannot marshal 'return value': Invalid managed/unmanaged type combination.
   at Gio.Internal.ApplicationCommandLine.GetArguments(IntPtr cmdline, Int32& argc)
   at Gio.ApplicationCommandLine.GetArguments(Int32& argc) in /home/user/.var/distrobox/fedora/Projects/gir.core/src/Libs/Gio-2.0/Public/ApplicationCommandLine.Methods.Generated.cs:line 30
   at org.gircore.test.Program.<>c.<.ctor>b__3_2(Application sender, CommandLineSignalArgs e) in /home/user/.var/distrobox/fedora/Projects/GirTest/Program.cs:line 29
   at GObject.ReturningSignal`3.<>c__DisplayClass11_0.<Connect>b__0(Value returnValue, Value[] parameters) in /home/user/.var/distrobox/fedora/Projects/gir.core/src/Libs/GObject-2.0/Public/ReturningSignalTSenderTSignalArgs.cs:line 50
   at GObject.Closure.InternalCallback(IntPtr closure, IntPtr returnValuePtr, UInt32 nParamValues, ValueData[] paramValuesData, IntPtr invocationHint, IntPtr userData) in /home/user/.var/distrobox/fedora/Projects/gir.core/src/Libs/GObject-2.0/Public/Closure.cs:line 47
```

**Comment by badcel** [comment]
Thanks for testing things out. I will take a look at your findings later.

Can you switch the samples to use `RunWithSynchrinizarionContext(null)` as it probably should be the recommended way of doing it to avoid problems for new users.

**Comment by badcel** [comment]
I could reproduce the problems with `HandleCommandLine` and added the defunct code as a sample to #955.

I could not reproduce the `HandlesOpen` issue but this is probably because I could not get the code to raise the event. From your exception it looks like an interface (`Gio.File`) should be extracted from signals args which is a `GObject.Value`. As extracting interfaces from a value is not yet supported I opened #958 for this. Could you provide some more help on how to get the sample working in issue #958?

As those points are not related to your PR it is good to go. Thank you :+1:

**Pull request merged**